### PR TITLE
Init serial read function (NOT MERGE)

### PR DESCRIPTION
**Pull request by sabas1080**
Hi

I'm trying to create the functions of read in the serial, but I have some problems in the part of typescript

```
Type '0' is not assignable to type 'string'.
shims.d.ts(11,14): error TS2384: Overload signatures must all be ambient or non-ambient.
shims.d.ts(19,14): error TS2384: Overload signatures must all be ambient or non-ambient.
```

Thanks

### CODE REVIEW

File: libs/serial/serial.cpp [hunk 1]
```diff
@@ -43,6 +58,47 @@ SINGLETON(WSerial);
...
 
 namespace serial {
+
+      /**
+     * Read a line of text from the serial port and return the buffer when the delimiter is met.
+     * @param delimiter text delimiter that separates each text chunk
+     */
+    //% help=serial/read-until
+    //% blockId=serial_read_until block="serial|read until %delimiter=serial_delimiter_conv"
+    //% weight=19
```
Comment (pelikhan) [inline libs/serial/serial.cpp]: move these annotations to the .ts file and leave a //%

File: libs/serial/sim/serial.ts [hunk 2]
```diff
@@ -1,4 +1,12 @@
 namespace pxsim.serial {
+    export function readUntil(delimiter: string) {
+
```
Comment (pelikhan) [inline libs/serial/sim/serial.ts]: return empty line at least

**Comment by sabas1080** [comment]
@pelikhan thanks for comments, I have made some changes but still I can not receive data

**Comment by pelikhan** [comment]
What is your test?

**Comment by sabas1080** [comment]
ok, the blink does not stop anymore, but I do not get anything

**Comment by pelikhan** [comment]
Did you try compiling a pure coral sample to see if the serial c++ is working?

**Comment by pelikhan** [comment]
Does it lock on readString?

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by sabas1080** [comment]
no, blink ok, but not read, apparently do not pass the data from C to typescript

**Comment by sabas1080** [comment]
my test is now

```
let dato = ""
serial.setBaudRate(BaudRate.BaudRate9600)
forever(function () {
    dato = serial.readString()
    console.log(dato)
    pins.LED.digitalWrite(true)
    pause(500)
    pins.LED.digitalWrite(false)
    pause(500)
})
```

makecode maker blocks javascript editor 1

**Comment by pelikhan** [comment]
Skip console, do serial.writeLine instead. I need to find the docs to attach the debugger.

**Comment by sabas1080** [comment]
but without console, no "console device", does it come from the device or not? and it works with webusb o no?

**Comment by pelikhan** [comment]
Console.log should work with WebUSB always.  You should add more logging to trace down where it breaks down. Is it not receiving any bytes?

**Comment by sabas1080** [comment]
I only receive that is in typescript

https://github.com/ElectronicCats/pxt-common-packages/blob/serial-read-test/libs/serial/serial.ts#L59

example if change "" for "a" receive "a"

**Comment by pelikhan** [comment]
Ha! You are missing shim=serial::readString !

**Comment by sabas1080** [comment]
Oh!

and if it is already in shims.d.s

https://github.com/ElectronicCats/pxt-common-packages/blob/serial-read-test/libs/serial/shims.d.ts#L18

no matter? Or what happens?

**Comment by pelikhan** [comment]
You should not need that one anymore - definitely wrong. Something is wrong with the notation. I think you need the shim notation on the c++ function comment.


Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

File: libs/serial/serial.ts [hunk 3]
```diff
@@ -37,5 +37,55 @@ namespace serial {
...
     }
+    
+    /**
+     * Read a line of text from the serial port and return the buffer when the delimiter is met.
+     * @param delimiter text delimiter that separates each text chunk
+     */
+    //% help=serial/read-until
+    //% blockId=serial_readUntil block="serial|read until %delimiter=serial_delimiter_conv"
+    //% weight=19
+    export function readUntil(del: string): string {
```
Comment (pelikhan) [inline libs/serial/serial.ts]: on a second though, you don't need the TypeScript implementatiuon fro this. Move the //% back to the C++ function and delete this function

File: libs/serial/serial.ts [hunk 4]
```diff
@@ -37,5 +37,55 @@ namespace serial {
...
+        return readString();
+    }
+    
+    /**
+    * Read the buffered received data as a string
+    */
+    //% help=serial/read-string
+    //% blockId=serial_read_buffer block="serial|read string"
+    //% weight=18
+    export function readString(): string {
```
Comment (pelikhan) [inline libs/serial/serial.ts]: same here, delete

File: libs/serial/serial.ts [hunk 5]
```diff
@@ -37,5 +37,55 @@ namespace serial {
...
+        return ""
+    }
 
+    /**
+    * Register an event to be fired when one of the delimiter is matched.
+    * @param delimiters the characters to match received characters against.
+    */
+    //% help=serial/on-data-received
+    //% weight=18 blockId=serial_on_data_received block="serial|on data received %delimiters=serial_delimiter_conv"
+    export function onDataReceived(delimiters: string, body: () => void): void{
```
Comment (pelikhan) [inline libs/serial/serial.ts]: same here delete

**Comment by pelikhan** [comment]
You will want to add function implementations in /libs/serial/sim/serial.ts

**Comment by sabas1080** [comment]
with the changes when starting the simulator

`warning: Program Error: Cannot read property 'vtable' of undefined`

**Comment by sabas1080** [comment]
@pelikhan 
I have managed to send C ++ information to typescript, but there is no reception

https://github.com/ElectronicCats/pxt-common-packages/blob/serial-read-test/libs/serial/serial.cpp#L73

but the condition is also not met

https://github.com/ElectronicCats/pxt-common-packages/blob/serial-read-test/libs/serial/serial.cpp#L72

File: libs/serial/serial.cpp [hunk 6]
```diff
@@ -69,10 +74,10 @@ namespace serial {
     //%
     String readString() {
       int n = getWSerial()->serial.getRxBufferSize();
-      if (n == 0) return mkString("", 0);
-      auto f = getWSerial()->serial.read(n, SerialMode::ASYNC);
-      auto res = mkString(f.toCharArray(), f.length());
-      return res;
+      if (n == 0) return mkString("null", 4);
```
Comment (pelikhan) [inline libs/serial/serial.cpp]: why?

Comment (sabas1080) [reply libs/serial/serial.cpp]: test just to know if it enters there

**Comment by sabas1080** [comment]
Blink for ok, but in read.String break and block of chips
My code example is
```
let da = ""
serial.attachToConsole()
serial.setBaudRate(BaudRate.BaudRate9600)
forever(function () {
    for (let i = 0; i < 4; i++) {
        pins.LED.digitalWrite(true)
        pause(500)
        pins.LED.digitalWrite(false)
        pause(500)
        pins.LED.digitalWrite(true)
        pause(500)
        pins.LED.digitalWrite(false)
        pause(500)
    }
    da = serial.readString()
})
```

File: libs/serial/serial.cpp (hunk 6 above)
Comment (sabas1080) [reply libs/serial/serial.cpp]: never enter this point

Comment (sabas1080) [reply libs/serial/serial.cpp]: buffer with "garbage"?

**Pull request closed without merging**